src/combinatorics/ramsey/infinite.lean

```diff
@@ -0,0 +1,483 @@
+/-
+Copyright (c) 2019 Jared Corduan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Jared Corduan.
+-/
+
+import logic.basic
+
+/-!
+# The Infinite Ramsey Theorem
+
+The main theorem, `infinite_ramsey_pairs_two_colors`, states that given
+a function from the unordered pairs of natural numbers to a set
+with two elements (often named `red` and `blue`), there exists an
+infinite set of natural numbers whose unordered pairs are all given
+the same color.
+
+The proof given here follows a very standard proof, such as the one
+given in chapter 1, theorem 5, of Graham, et al.
+
+The proof roughly goes as follows:
+
+Given a function `f` of unordered pairs into the colors `red` and `blue`,
+define a sequence `(Sᵢ, xᵢ)` so that:
+
+* (S₀, x₀) = (ℕ, 0)
+* Sᵢ₊₁ ⊆ Sᵢ
+* xᵢ₊₁ ∈ Sᵢ
+* xᵢ < xᵢ₊₁
+* f {xᵢ₊₁, y} = f {xᵢ₊₁, z} for all xᵢ₊₁ < y < z in Sₓ, x = xᵢ
+
+Then the set {xᵢ | i ∈ ℕ} has the property that
+f {xᵢ, xⱼ} = f {xᵢ, xₖ} for all i < j < k.
+If we color singletons in this set by f' xᵢ = f {xᵢ, xᵢ₊₁},
+then by the pigeon hole principle there is an infinite subset
+of the xᵢ's whose unordered pairs are all colored the same.
+
+## Notation
+
+This file uses `[S]²` to denote the set of unordered pairs of
+the given set `S`.
+
+## Implementation notes
+
+The set of two colors is defined as an emumeration.
+This definition is easy to work with, and easy to read,
+but comes at the cost of not generalizing.
+
+Unordered pairs are implemented as ordered pairs with the condition
+that the second element is strictly greater than the first.
+This definition is also easy to work with but does not generalize.
+
+## References
+
+*  [Graham, R.L. and Rothschild, B.L. and Spencer, J.H., *Ramsey Theory*][graham1990]
+-/
+
+open set
+open classical
+open nat
+
+section infinite_ramsey_pairs
+
+/--
+A set of natural numbers H is infinite if for any natural
+number there is a larger number in H.
+-/
+def infinite (H : set ℕ) := ∀ x : ℕ, ∃ y : ℕ, x < y ∧ y ∈ H
+
+/--
+There are two colors, red and blue.
+-/
+inductive color
```

Comment:
We don't want this definition to end up in the root namespace, so it would be good to turn `section infinite_ramsey_pairs` into `namespace infinite_ramsey_pairs`.

Reply:
when I use a namespace instead of a section, I see that I need someway of separating my instances. For example, I get an error on my second `has_coe`, saying "inifinite_ramsey_pairs.has_coe has already been defined". Is the recommended solution to put things in sections inside of the namespace?

Reply:
This is an annoying Lean bug--it generates bad names for instances when inside a namespace. (I don't know if also being inside a section helps, does it?) In any case you can just assign the names yourself, like `instance Inf.has_coe : has_coe Inf ...`.

Reply:
ah, that makes sense, thank you for the explanation!